core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala

```diff
@@ -0,0 +1,360 @@
+              // Turn FileStatus into SerializableFileStatus so we can send it back to the driver
+              val blockLocations = status match {
+                case f: LocatedFileStatus =>
+                  f.getBlockLocations.map { loc =>
+                    SerializableBlockLocation(
+                      loc.getNames,
+                      loc.getHosts,
+                      loc.getOffset,
+                      loc.getLength)
+                  }
+
+                case _ =>
+                  Array.empty[SerializableBlockLocation]
+              }
+
+              SerializableFileStatus(
+                status.getPath.toString,
+                status.getLen,
+                status.isDirectory,
+                status.getReplication,
+                status.getBlockSize,
+                status.getModificationTime,
+                status.getAccessTime,
+                blockLocations)
+            }
+            (path.toString, serializableStatuses)
+        }.collect()
+    } finally {
+      sc.setJobDescription(previousJobDescription)
+    }
+
+    // turn SerializableFileStatus back to Status
+    statusMap.map { case (path, serializableStatuses) =>
+      val statuses = serializableStatuses.map { f =>
+        val blockLocations = f.blockLocations.map { loc =>
+          new BlockLocation(loc.names, loc.hosts, loc.offset, loc.length)
+        }
+        new LocatedFileStatus(
+          new FileStatus(
+            f.length, f.isDir, f.blockReplication, f.blockSize, f.modificationTime,
+            new Path(f.path)),
+          blockLocations)
+      }
+      (new Path(path), statuses)
+    }
+  }
+
+  // scalastyle:off argcount
+  /**
+   * Lists a single filesystem path recursively. If a `SparkContext`` object is specified, this
+   * function may launch Spark jobs to parallelize listing based on parallelismThreshold.
+   *
+   * If sessionOpt is None, this may be called on executors.
+   *
+   * @return all children of path that match the specified filter.
+   */
+  private def listLeafFiles(
+      path: Path,
+      hadoopConf: Configuration,
+      filter: PathFilter,
+      contextOpt: Option[SparkContext],
+      ignoreMissingFiles: Boolean,
+      ignoreLocality: Boolean,
+      isSQLRootPath: Boolean,
+      filterFun: Option[String => Boolean],
+      parallelismThreshold: Int,
+      parallelismDefault: Int): Seq[FileStatus] = {
+
+    logTrace(s"Listing $path")
+    val fs = path.getFileSystem(hadoopConf)
+
+    // Note that statuses only include FileStatus for the files and dirs directly under path,
+    // and does not include anything else recursively.
+    val statuses: Array[FileStatus] = try {
+      fs match {
+        // DistributedFileSystem overrides listLocatedStatus to make 1 single call to namenode
+        // to retrieve the file status with the file block location. The reason to still fallback
+        // to listStatus is because the default implementation would potentially throw a
```

Comment:
I would just switch to the incremental one *everywhere*

HDFS likes it because when you have many, many files in a path they can release the lock on the NN; for object stores they always have to page in (s3, azure, GCS)....which gives the implementors an opportunity to move from the default implementation to exposing the incremental one. Add this and we can just let the relevant teams know.

w.r.t CommonPathCapabilities, I suppose we could add one which declares that the listing ops are paged. But do you really want code to try and be that clever? I'm trying to use that feature to mark up

* where optional features will fail with UnimplementedException or similar without you having to try using them (append, truncate, xattrs)
* where there are fundamental semantics its important for algorithms to worry about (hflush)
* where your store wants to expose some state to callers (e.g. s3a exposing its HADOOP-13230 dir marker policy)

If you want a performance option, we could add one.

BTW, PathCapabilities is in hadoop-3.2.x now, will be in next release.  I might do it for 3.1 too...it makes a good way to programmatically/CLI probe for s3a dir marker policy, see.

Also, cloudstore has some CLI commands for the list calls (and pathcapabilities) to help explore what's going on, including, on s3a, listing cost in #of HTTP requests. Worth playing with to see what is good/bad, though as Mukund has been doing lots of work on 3.3.x s3a list, it will look worse than you'd expect on a dir treewalk,
https://github.com/steveloughran/cloudstore

Reply:
I think changing this in a follow up PR sounds fine to me. I'd like to us to use the faster method by default and fall back on exception, but that could be a follow on. Want to file a JIRA for broadening the scope of using the faster method?

Reply:
Thanks @steveloughran !

> I would just switch to the incremental one everywhere

Yes agree. This will help to simplify the code here a lot, although I guess we need to make sure this is enforced on the Hadoop side and that any future derived `FileSystem` implementations won't inherit the default behavior.

Back to the code, seems the main reason why it calls `listStatus` in the else case (e.g., not `DistributedFileSystem`) is because it want to save `getBlockLocations` cost by calling that later on filtered leaf files. We can potentially use `FilterFileSystem` here and pass the filter to it instead. I'll try to explore this in a follow-up PR.

> Want to file a JIRA for broadening the scope of using the faster method?

Yes will do that. The PR will be about improving and cleanup this utility class.

Reply:
w.r.t the faster (incremental) calls, yes, something to consider next. At the very least, you will be able to collect and report/aggregate stats. For example, here is me getting the stats on a large/deep list just by calling toString on the RemoteIterator after it's done its work

```
Listing statistics:
  counters=((object_list_request=1) (object_continue_list_request=40)); gauges=(); minimums=((object_list_request.min=1298) (object_continue_list_request.min=414)); maximums=((object_continue_list_request.max=746) (object_list_request.max=1298)); means=((object_continue_list_request.mean=(sum=18210, samples=39, mean=466.9231)) (object_list_request.mean=(sum=1298, samples=1, mean=1298.0000))); 

2020-08-27 13:25:21,122 [main] INFO  tools.MarkerTool (DurationInfo.java:close(98)) - marker scan s3a://landsat-pds/: duration 0:19.563s
Listed 40000 objects under s3a://landsat-pds/
```
Now, who wouldn't like to know things like that. And ideally, collect across threads and merge back in.

As an aside, looked at {{org.apache.hadoop.mapred.LocatedFileStatusFetcher}}. This does multithreaded status fetching and collects those stats. Although it's tagged Private, I've noticed Parquet uses it so my next PR will convert to public/evolving and document the fact. 

If you could use that, we could look @ evolving it better, especially returning a RemoteIterator of results which we could incrementally fill in across threads rather than block for the final results. Anything which makes it possible for app code to process data (read footers, etc) while the listing goes on has significant benefit in the World of Object Stores

Reply:
Yes, stats will be nice. Looking forward to that 👍 (is it tracked by [this PR](https://github.com/apache/hadoop/pull/2069)?)

w.r.t `LocatedFileStatusFetcher`, yes it will be great if this is exposed as a public API. I think we should also consider moving this to another module such as hadoop-common, and perhaps adding an option to turn on/off locality (so that we don't have to get the block locations if they are not needed). Pairing it with batch listing can also potentially help (although this is only available in HDFS currently).

Instead of multiple threads, Spark currently use a distributed job so I'm not sure whether there will be any regression by doing that, but we can explore this later.

Reply:
> w.r.t LocatedFileStatusFetcher, yes it will be great if this is exposed as a public API. I think we should also consider moving this to another module such as hadoop-common, and perhaps adding an option to turn on/off locality (so that we don't have to get the block locations if they are not needed). Pairing it with batch listing can also potentially help (although this is only available in HDFS currently).

I'd worry about the effect of moving the package as is. It'd have to be a copy and paste, or move and subclass

Reply:
@sunchao -yes, that's the stats API. Now is the time to review it and tell me where it doesn't suit your (perceived future) needs. I've been playing with a Parquet branch which uses it, not in Spark itself.

Reply:
Thanks for the info @steveloughran . I'll check out the PR (it is a very big one though).